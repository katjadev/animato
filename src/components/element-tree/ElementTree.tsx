import { FC, ReactNode, useEffect, useState } from 'react'
import { useTranslations } from 'next-intl'
import { v4 as uuidv4 } from 'uuid'
import Icon from '@animato/components/icon/Icon';
import styles from './ElementTree.module.css'

type ElementTreeNode = {
  id: string,
  title: string,
  element: Element,
  children: ElementTreeNode[],
}

const ALLOWED_NODES = ['g', 'path', 'rect', 'circle']

interface ElementTreeProps {
  projectId: string;
  content: string;
}

const ElementTree: FC<ElementTreeProps> = ({
  projectId,
  content,
}) => {
  const t = useTranslations('project')
  const [elements, setElements] = useState<ElementTreeNode[]>([])
  const [collapsedNodes, setCollapsedNodes] = useState<string[]>(JSON.parse(localStorage.getItem(`${projectId}-collapsed-elements`) || '[]'))

  useEffect(() => {
    const parser = new DOMParser()
    const doc = parser.parseFromString(content, 'application/xml')
    
    const makeTree = (currentElement: Element): ElementTreeNode[] => {
      const children = Array.from(currentElement.children)
        .filter((element) => ALLOWED_NODES.includes(element.nodeName))

      return children.map((element) => ({
        id: element.getAttribute('id') || uuidv4(),
        element,
        title: element.getAttribute('data-title') || element.nodeName,
        children: makeTree(element),
      }))
    }

    const tree = makeTree(doc.documentElement)
    setElements(tree)
  }, [content])
  
  useEffect(() => {
    localStorage.setItem(`${projectId}-collapsed-elements`, JSON.stringify(collapsedNodes))
  }, [collapsedNodes])

  const toggleNode = (elementId: string): void => {
    if (collapsedNodes.includes(elementId)) {
      setCollapsedNodes(collapsedNodes.filter((id) => id !== elementId))
    } else {
      setCollapsedNodes([...collapsedNodes, elementId])
    }
  }

  const renderTree = (list: ElementTreeNode[]): ReactNode => (
    <>
      {list.map((element: ElementTreeNode) => (
        <div 
          key={element.id} 
          className={styles.treeNode}
        >
          <div className={`${styles.nodeTitle} ${!element.children.length ? styles.lastChild : ''}`}>
            {element.children.length > 0 && (
              <button
                className={styles.collapseButton}
                onClick={() => toggleNode(element.id)}
              >
                <Icon icon={collapsedNodes.includes(element.id) ? 'icon-arrow_right' : 'icon-arrow_drop_down'} />
              </button>
            )}
            <div>{element.title}</div>
          </div>
          {element.children.length > 0 && !collapsedNodes.includes(element.id) && (
            <>{renderTree(element.children)}</>
          )}
        </div>
      ))}
    </>
  )

  return (
    <div className={styles.elements}>
      <h2 className={styles.title}>{t('elements')}</h2>
      {renderTree(elements)}
    </div>
  )
}

export default ElementTree