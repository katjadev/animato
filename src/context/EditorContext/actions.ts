import { Dispatch } from 'react'
import { Action } from '@animato/types'

export const actionTypes = {
  START_PLAYING: 'START_PLAYING',
  STOP_PLAYING: 'STOP_PLAYING',
  TOGGLE_REPEAT_MODE: 'TOGGLE_REPEAT_MODE',
  HOVER_ELEMENT: 'HOVER_ELEMENT',
  TOGGLE_ELEMENT: 'TOGGLE_ELEMENT',
  SET_CURRENT_TIME: 'SET_CURRENT_TIME',
  SET_TIMELINE_WIDTH: 'SET_TIMELINE_WIDTH',
  SET_SCROLL_POSITION: 'SET_SCROLL_POSITION',
  SET_ZOOM: 'SET_ZOOM',
  COLLAPSE_ELEMENT: 'COLLAPSE_ELEMENT',
  EXPAND_ELEMENT: 'EXPAND_ELEMENT',
  COLLAPSE_ANIMATION: 'COLLAPSE_ANIMATION',
  EXPAND_ANIMATION: 'EXPAND_ANIMATION',
}

export const editorActions = (dispatch: Dispatch<Action>) => (
  {
    startPlaying: () => dispatch({ type: actionTypes.START_PLAYING }),
    stopPlaying: () => dispatch({ type: actionTypes.STOP_PLAYING }),
    toggleRepeatMode: () => dispatch({ type: actionTypes.TOGGLE_REPEAT_MODE }),
    hoverElement: (payload: { id: string | null }) => dispatch({ type: actionTypes.HOVER_ELEMENT, payload }),
    toggleElement: (payload: { id: string, multiple?: boolean }) => dispatch({ type: actionTypes.TOGGLE_ELEMENT, payload }),
    setCurrentTime: (payload: { value: number }) => dispatch({ type: actionTypes.SET_CURRENT_TIME, payload }),
    setTimelineWidth: (payload: { width: number }) => dispatch({ type: actionTypes.SET_TIMELINE_WIDTH, payload }),
    setScrollPosition: (payload: { value: number }) => dispatch({ type: actionTypes.SET_SCROLL_POSITION, payload }),
    setZoom: (payload: { value: number }) => dispatch({ type: actionTypes.SET_ZOOM, payload }),
    collapseElement: (payload: { id: string }) => dispatch({ type: actionTypes.COLLAPSE_ELEMENT, payload }),
    expandElement: (payload: { id: string }) => dispatch({ type: actionTypes.EXPAND_ELEMENT, payload }),
    collapseAnimation: (payload: { id: string }) => dispatch({ type: actionTypes.COLLAPSE_ANIMATION, payload }),
    expandAnimation: (payload: { id: string }) => dispatch({ type: actionTypes.EXPAND_ANIMATION, payload }),
  }
)
