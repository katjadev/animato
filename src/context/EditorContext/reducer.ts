import { Action, ScrollPosition, TimelineMark } from '@animato/types'
import { MAX_ZOOM, REM_TO_PX_COEFFICIENT, TIMELINE_PADDING } from '@animato/constants'
import generateTimelineMarks from '@animato/utils/generateTimelineMarks'
import { actionTypes } from './actions'

export type EditorState = {
  isPlaying: boolean;
  isRepeatMode: boolean;
  hoveredElementId: string | null;
  selectedElementIds: string[];
  currentTime: number;
  timelineWidth: number;
  scrollPosition: ScrollPosition;
  zoom: number;
  timelineMarks: TimelineMark[];
  collapsedElements: string[];
  collapsedAnimations: string[];
}

const timelinePaddingPx = TIMELINE_PADDING * REM_TO_PX_COEFFICIENT
const timelineMarks = generateTimelineMarks(MAX_ZOOM)
const markSize = timelineMarks[1].position - timelineMarks[0].position
const timelineWidth = timelineMarks.length * markSize + 2 * timelinePaddingPx

export const initialEditorState: EditorState = {
  isPlaying: false,
  isRepeatMode: false,
  hoveredElementId: null,
  selectedElementIds: [],
  currentTime: 0,
  timelineWidth,
  scrollPosition: { top: 0, left: 0 },
  zoom: MAX_ZOOM,
  timelineMarks,
  collapsedElements: [],
  collapsedAnimations: [],
}

export const editorReducer = (state: EditorState, action: Action) => {
  switch (action.type) {
    case actionTypes.START_PLAYING:
      return {
        ...state,
        isPlaying: true,
      }
    case actionTypes.STOP_PLAYING:
      return {
        ...state,
        isPlaying: false,
      }
    case actionTypes.TOGGLE_REPEAT_MODE:
      return {
        ...state,
        isRepeatMode: !state.isRepeatMode,
      }
    case actionTypes.HOVER_ELEMENT:
      return {
        ...state,
        hoveredElementId: action.payload!.id,
      }
    case actionTypes.SELECT_ELEMENT:
      return {
        ...state,
        selectedElementIds: [...state.selectedElementIds, action.payload!.id],
      }
    case actionTypes.DESELECT_ELEMENT:
      return {
        ...state,
        selectedElementIds: state.selectedElementIds.filter(id => id !== action.payload!.id),
      }
    case actionTypes.SET_CURRENT_TIME:
      return {
        ...state,
        currentTime: action.payload!.value,
      }
    case actionTypes.SET_TIMELINE_WIDTH:
      return {
        ...state,
        timelineWidth: action.payload!.width,
      }
    case actionTypes.SET_SCROLL_POSITION:
      return {
        ...state,
        scrollPosition: action.payload!.value,
      }
    case actionTypes.SET_ZOOM:
      const zoom = action.payload!.value
      const timelineMarks = generateTimelineMarks(zoom)
      const markSize = timelineMarks[1].position - timelineMarks[0].position
      const timelineWidth = timelineMarks.length * markSize + 2 * timelinePaddingPx
      return {
        ...state,
        zoom,
        timelineMarks,
        timelineWidth,
      }
    case actionTypes.COLLAPSE_ELEMENT:
      return {
        ...state,
        collapsedElements: [...state.collapsedElements, action.payload!.id],
      }
    case actionTypes.EXPAND_ELEMENT:
      return {
        ...state,
        collapsedElements: state.collapsedElements.filter(id => id !== action.payload!.id),
      }
    case actionTypes.COLLAPSE_ANIMATION:
      return {
        ...state,
        collapsedAnimations: [...state.collapsedAnimations, action.payload!.id],
      }
    case actionTypes.EXPAND_ANIMATION:
      return {
        ...state,
        collapsedAnimations: state.collapsedAnimations.filter(id => id !== action.payload!.id),
      }
    default:
      return state
  }
}
