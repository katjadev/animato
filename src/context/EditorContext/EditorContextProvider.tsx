import { ReactNode, createContext, useContext, useReducer } from 'react'
import { EditorState, editorReducer, initialEditorState } from './reducer'
import { editorActions } from './actions'



export const EditorContext = createContext<{
  state: EditorState;
  actions: any;
}>({
  state: initialEditorState,
  actions: {},
})

interface EditorContextProviderProps {
  children: ReactNode;
}

export default function EditorContextProvider({children}: EditorContextProviderProps) {
  const [state, dispatch] = useReducer(editorReducer, initialEditorState)
  const actions = editorActions(dispatch)

  return (
    <EditorContext.Provider value={{ state, actions }}>
      {children}
    </EditorContext.Provider>
  )
}

export const useEditorState = () => useContext(EditorContext)