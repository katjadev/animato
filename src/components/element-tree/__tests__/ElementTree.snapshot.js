import React from 'react'
import { render } from '@testing-library/react'
import { useEditorState } from '@animato/context/EditorContext/EditorContextProvider'
import ElementTree from '../ElementTree'

jest.mock('@animato/context/EditorContext/EditorContextProvider', () => ({
  useEditorState: jest.fn(() => ({
    state: { collapsedElements: [], selectedElementIds: [] },
  })),
}))

const TEST_ID = 'test-id'

jest.mock('@animato/context/ProjectContext/ProjectContextProvider', () => ({
  useProjectState: jest.fn(() => ({
    state: { elements: [
      {
        id: 'test-id',
        tagName: 'rect',
        title: 'Element 1',
        children: [
          {
            id: 'element2',
            tagName: 'path',
            title: 'Element 2',
            children: [],
          },
        ],
      },
    ] },
  })),
}))

const translations = {
  expand: 'Expand',
  collapse: 'Collapse',
  elements: 'Elements',
}

describe('ElementTree', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ElementTree translations={translations} />
    )

    expect(container).toMatchSnapshot()
  })

  it('renders correctly with collapsed elements', () => {
    useEditorState.mockReturnValue({
      state: { collapsedElements: [TEST_ID], selectedElementIds: [] },
      actions: {},
    })

    const { container } = render(
      <ElementTree translations={translations} />
    )

    expect(container).toMatchSnapshot()
  })
})
