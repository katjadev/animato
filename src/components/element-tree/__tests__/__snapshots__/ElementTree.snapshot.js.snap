// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementTree renders correctly 1`] = `
<div>
  <div
    class=" elements"
  >
    <h2
      class="title"
    >
      Elements
    </h2>
    <div
      class="treeNode"
    >
      <div
        class="title  "
        tabindex="0"
      >
        <div
          class="titleIn level0"
        >
          <button
            aria-label="Collapse"
            class="collapseButton"
          >
            <i
              class="iconoir-nav-arrow-down "
            />
          </button>
          <i
            class="iconoir-square shapeIcon"
          />
          <div
            class="titleText"
          >
            <span
              class="text"
              tabindex="0"
            >
              Element 1
            </span>
          </div>
          <button
            class="iconButton  small"
          >
            <i
              class="iconoir-more-vert "
            />
          </button>
        </div>
      </div>
      <div
        class="treeNode"
      >
        <div
          class="title  "
          tabindex="0"
        >
          <div
            class="titleIn level1"
          >
            <i
              class="iconoir-3d-bridge shapeIcon"
            />
            <div
              class="titleText"
            >
              <span
                class="text"
                tabindex="0"
              >
                Element 2
              </span>
            </div>
            <button
              class="iconButton  small"
            >
              <i
                class="iconoir-more-vert "
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementTree renders correctly with collapsed elements 1`] = `
<div>
  <div
    class=" elements"
  >
    <h2
      class="title"
    >
      Elements
    </h2>
    <div
      class="treeNode"
    >
      <div
        class="title  "
        tabindex="0"
      >
        <div
          class="titleIn level0"
        >
          <button
            aria-label="Expand"
            class="collapseButton"
          >
            <i
              class="iconoir-nav-arrow-right "
            />
          </button>
          <i
            class="iconoir-square shapeIcon"
          />
          <div
            class="titleText"
          >
            <span
              class="text"
              tabindex="0"
            >
              Element 1
            </span>
          </div>
          <button
            class="iconButton  small"
          >
            <i
              class="iconoir-more-vert "
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
